File: frontend/lib/types/learning.ts
import type { EconomyAction } from "./economy";

export type LessonListItem = {
  id: string;
  slug: string;
  title: string;
  min_tier: string;
  sort_order: number;
  created_at: string;
  locked: boolean;
};

export type PopularLessonItem = LessonListItem & {
  completion_count: number;
};

export type LessonDetail = LessonListItem & {
  body: string;
  body_locked?: boolean;
};

export type LearningProgressStatus = "not_started" | "active" | "completed";
export type LearningLessonStatus = "not_started" | "in_progress" | "completed";
export type LearningStepKind =
  | "theory"
  | "guided_practice"
  | "quiz"
  | "applied_exercise"
  | "reflection"
  | "final_checkpoint";

export type LearningStartTarget = {
  target: string;
  has_active_course: boolean;
  active_course_slug?: string | null;
  resume_href?: string | null;
};

export type LearningActionLink = {
  label: string;
  href: string;
  body?: string | null;
};

export type LearningCourseCard = {
  slug: string;
  title: string;
  subtitle: string;
  description: string;
  difficulty: string;
  result_headline?: string | null;
  deliverable_preview?: string | null;
  estimated_minutes: number;
  module_count: number;
  lesson_count: number;
  progress_percent: number;
  status: LearningProgressStatus;
  last_activity_at?: string | null;
  next_lesson_slug?: string | null;
  resume_href?: string | null;
  badge_earned: boolean;
  course_reward_lmn: number;
};

export type LearningCatalog = {
  courses: LearningCourseCard[];
  recommended_course_slug?: string | null;
};

export type LearningWeakArea = {
  tag: string;
  count: number;
  recommendation: string;
  lesson_slug?: string | null;
};

export type LearningMyCourseItem = {
  slug: string;
  title: string;
  subtitle: string;
  progress_percent: number;
  status: LearningProgressStatus;
  last_activity_at?: string | null;
  next_lesson_title?: string | null;
  next_lesson_slug?: string | null;
  continue_href?: string | null;
  completed_at?: string | null;
  badge_code?: string | null;
  certificate_ready: boolean;
};

export type LearningMyModules = {
  active_courses: LearningMyCourseItem[];
  completed_courses: LearningMyCourseItem[];
  weak_areas: LearningWeakArea[];
};

export type LearningLessonOutline = {
  slug: string;
  title: string;
  summary: string;
  estimated_minutes: number;
  position: number;
  status: LearningLessonStatus;
  unlocked: boolean;
  is_final_assessment: boolean;
  progress_percent: number;
  continue_href: string;
};

export type LearningModule = {
  slug: string;
  title: string;
  summary: string;
  position: number;
  lesson_count: number;
  progress_percent: number;
  lessons: LearningLessonOutline[];
};

export type LearningCourseRewards = {
  lesson_reward_lmn: number;
  course_reward_lmn: number;
  badge_code: string;
  certificate_template: string;
  badge_earned: boolean;
  course_completed: boolean;
};

export type LearningCourseDetail = {
  slug: string;
  title: string;
  subtitle: string;
  description: string;
  difficulty: string;
  result_headline?: string | null;
  estimated_minutes: number;
  module_count: number;
  lesson_count: number;
  progress_percent: number;
  status: LearningProgressStatus;
  last_activity_at?: string | null;
  resume_href?: string | null;
  start_or_continue_label: string;
  what_you_will_learn: string[];
  prerequisites: string[];
  deliverables: string[];
  career_outcomes: string[];
  product_action?: LearningActionLink | null;
  modules: LearningModule[];
  rewards: LearningCourseRewards;
  weak_areas: LearningWeakArea[];
};

export type LearningStepChoice = {
  id: string;
  text: string;
  explanation?: string | null;
};

export type LearningQuizQuestion = {
  id: string;
  question: string;
  choices: LearningStepChoice[];
};

export type LearningStepFeedback = {
  verdict: string;
  score: number;
  pass_score: number;
  strengths: string[];
  improvements: string[];
  revisit: string[];
  hint?: string | null;
};

export type LearningLessonStep = {
  slug: string;
  kind: LearningStepKind;
  title: string;
  estimated_minutes: number;
  content: string[];
  task?: string | null;
  placeholder?: string | null;
  question?: string | null;
  choices: LearningStepChoice[];
  quiz_questions: LearningQuizQuestion[];
  pass_score: number;
  min_words?: number | null;
  required_markers: string[];
  bonus_markers: string[];
  forbidden_phrases: string[];
  submission_type: "none" | "text" | "choice";
  unlocked: boolean;
  completed: boolean;
  attempts: number;
  last_score?: number | null;
  last_answer_text?: string | null;
  last_choice_id?: string | null;
  last_choice_map: Record<string, string>;
  feedback?: LearningStepFeedback | null;
};

export type LearningLessonDetail = {
  course_slug: string;
  module_slug: string;
  lesson_slug: string;
  title: string;
  summary: string;
  objective?: string | null;
  deliverable?: string | null;
  scenario_title?: string | null;
  scenario_body?: string | null;
  debrief: string[];
  review_rubric: string[];
  common_mistakes: string[];
  estimated_minutes: number;
  position_in_course: number;
  total_lessons: number;
  progress_percent: number;
  course_progress_percent: number;
  status: LearningLessonStatus;
  unlocked: boolean;
  is_final_assessment: boolean;
  return_to_course_href: string;
  previous_lesson_href?: string | null;
  next_lesson_href?: string | null;
  steps: LearningLessonStep[];
  current_step_slug?: string | null;
  lesson_list: LearningLessonOutline[];
};

export type LearningStepSubmitResponse = {
  course_slug: string;
  module_slug: string;
  lesson_slug: string;
  step_slug: string;
  passed: boolean;
  completed: boolean;
  score: number;
  attempts: number;
  feedback: LearningStepFeedback;
  lesson_progress_percent: number;
  course_progress_percent: number;
  lesson_completed: boolean;
  course_completed: boolean;
  next_step_slug?: string | null;
  next_lesson_slug?: string | null;
  resume_href: string;
  weak_areas: LearningWeakArea[];
  awarded_lmn: number;
  awarded_badge?: string | null;
  certificate_ready: boolean;
  economy?: EconomyAction | null;
};

File: backend/app/modules/learning/model/learning.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.modules.economy.model.store import EconomyActionRead

LearningStepKind = Literal[
    "theory",
    "guided_practice",
    "quiz",
    "applied_exercise",
    "reflection",
    "final_checkpoint",
]
LearningProgressStatus = Literal["not_started", "active", "completed"]
LearningLessonStatus = Literal["not_started", "in_progress", "completed"]


class LearningStartTargetRead(BaseModel):
    target: str
    has_active_course: bool
    active_course_slug: str | None = None
    resume_href: str | None = None


class LearningActionLinkRead(BaseModel):
    label: str
    href: str
    body: str | None = None


class LearningCourseCardRead(BaseModel):
    slug: str
    title: str
    subtitle: str
    description: str
    difficulty: str
    result_headline: str | None = None
    deliverable_preview: str | None = None
    estimated_minutes: int
    module_count: int
    lesson_count: int
    progress_percent: int = 0
    status: LearningProgressStatus = "not_started"
    last_activity_at: datetime | None = None
    next_lesson_slug: str | None = None
    resume_href: str | None = None
    badge_earned: bool = False
    course_reward_lmn: int = 0


class LearningCatalogRead(BaseModel):
    courses: list[LearningCourseCardRead]
    recommended_course_slug: str | None = None


class LearningMyCourseItemRead(BaseModel):
    slug: str
    title: str
    subtitle: str
    progress_percent: int
    status: LearningProgressStatus
    last_activity_at: datetime | None = None
    next_lesson_title: str | None = None
    next_lesson_slug: str | None = None
    continue_href: str | None = None
    completed_at: datetime | None = None
    badge_code: str | None = None
    certificate_ready: bool = False


class LearningWeakAreaRead(BaseModel):
    tag: str
    count: int
    recommendation: str
    lesson_slug: str | None = None


class LearningMyModulesRead(BaseModel):
    active_courses: list[LearningMyCourseItemRead]
    completed_courses: list[LearningMyCourseItemRead]
    weak_areas: list[LearningWeakAreaRead] = Field(default_factory=list)


class LearningLessonOutlineRead(BaseModel):
    slug: str
    title: str
    summary: str
    estimated_minutes: int
    position: int
    status: LearningLessonStatus
    unlocked: bool
    is_final_assessment: bool
    progress_percent: int
    continue_href: str


class LearningModuleRead(BaseModel):
    slug: str
    title: str
    summary: str
    position: int
    lesson_count: int
    progress_percent: int
    lessons: list[LearningLessonOutlineRead]


class LearningCourseRewardsRead(BaseModel):
    lesson_reward_lmn: int
    course_reward_lmn: int
    badge_code: str
    certificate_template: str
    badge_earned: bool = False
    course_completed: bool = False


class LearningCourseRead(BaseModel):
    slug: str
    title: str
    subtitle: str
    description: str
    difficulty: str
    result_headline: str | None = None
    estimated_minutes: int
    module_count: int
    lesson_count: int
    progress_percent: int
    status: LearningProgressStatus
    last_activity_at: datetime | None = None
    resume_href: str | None = None
    start_or_continue_label: str
    what_you_will_learn: list[str]
    prerequisites: list[str] = Field(default_factory=list)
    deliverables: list[str] = Field(default_factory=list)
    career_outcomes: list[str] = Field(default_factory=list)
    product_action: LearningActionLinkRead | None = None
    modules: list[LearningModuleRead]
    rewards: LearningCourseRewardsRead
    weak_areas: list[LearningWeakAreaRead] = Field(default_factory=list)


class LearningStepChoiceRead(BaseModel):
    id: str
    text: str
    explanation: str | None = None


class LearningQuizQuestionRead(BaseModel):
    id: str
    question: str
    choices: list[LearningStepChoiceRead] = Field(default_factory=list)


class LearningStepFeedbackRead(BaseModel):
    verdict: str
    score: int
    pass_score: int
    strengths: list[str] = Field(default_factory=list)
    improvements: list[str] = Field(default_factory=list)
    revisit: list[str] = Field(default_factory=list)
    hint: str | None = None


class LearningLessonStepRead(BaseModel):
    slug: str
    kind: LearningStepKind
    title: str
    estimated_minutes: int
    content: list[str] = Field(default_factory=list)
    task: str | None = None
    placeholder: str | None = None
    question: str | None = None
    choices: list[LearningStepChoiceRead] = Field(default_factory=list)
    quiz_questions: list[LearningQuizQuestionRead] = Field(default_factory=list)
    pass_score: int = 0
    min_words: int | None = None
    required_markers: list[str] = Field(default_factory=list)
    bonus_markers: list[str] = Field(default_factory=list)
    forbidden_phrases: list[str] = Field(default_factory=list)
    submission_type: Literal["none", "text", "choice"] = "none"
    unlocked: bool = True
    completed: bool = False
    attempts: int = 0
    last_score: int | None = None
    last_answer_text: str | None = None
    last_choice_id: str | None = None
    last_choice_map: dict[str, str] = Field(default_factory=dict)
    feedback: LearningStepFeedbackRead | None = None


class LearningLessonRead(BaseModel):
    course_slug: str
    module_slug: str
    lesson_slug: str
    title: str
    summary: str
    objective: str | None = None
    deliverable: str | None = None
    scenario_title: str | None = None
    scenario_body: str | None = None
    debrief: list[str] = Field(default_factory=list)
    review_rubric: list[str] = Field(default_factory=list)
    common_mistakes: list[str] = Field(default_factory=list)
    estimated_minutes: int
    position_in_course: int
    total_lessons: int
    progress_percent: int
    course_progress_percent: int
    status: LearningLessonStatus
    unlocked: bool
    is_final_assessment: bool
    return_to_course_href: str
    previous_lesson_href: str | None = None
    next_lesson_href: str | None = None
    steps: list[LearningLessonStepRead]
    current_step_slug: str | None = None
    lesson_list: list[LearningLessonOutlineRead]


class LearningStepSubmitRequest(BaseModel):
    answer: dict[str, Any] | None = None


class LearningStepSubmitRead(BaseModel):
    course_slug: str
    module_slug: str
    lesson_slug: str
    step_slug: str
    passed: bool
    completed: bool
    score: int
    attempts: int
    feedback: LearningStepFeedbackRead
    lesson_progress_percent: int
    course_progress_percent: int
    lesson_completed: bool
    course_completed: bool
    next_step_slug: str | None = None
    next_lesson_slug: str | None = None
    resume_href: str
    weak_areas: list[LearningWeakAreaRead] = Field(default_factory=list)
    awarded_lmn: int = 0
    awarded_badge: str | None = None
    certificate_ready: bool = False
    economy: EconomyActionRead | None = None

